### PR TITLE
Concurrency helpers

### PR DESCRIPTION
**Pull request by fh-ms**
New helpers for concurrency handling using ReentrantReadWriteLocks internally.

`LockedExecutor` and `StripeLockedExecutor` are utilities that use ReentrantReadWriteLock internally.

`LockScope` and `StripeLockScope` are abstract classes to extend. They provide protected delegates for the lock executors but hide the executor itself for even easier usage.

Use executors
```java
public class Customers
{
	private final transient LockedExecutor executor = LockedExecutor.New();
	private final List<Customer> customers = new ArrayList<>();
	
	public void addCustomer(Customer c)
	{
		this.executor.write(() -> {
			this.customers.add(c);
			Application.storageManager().store(this.customers);
		});
	}
	
	public void traverseCustomers(Consumer<Customer> consumer)
	{
		this.executor.read(() -> this.customers.forEach(consumer));
	}
}
```

Or extend from a scope
```java
public class Customers extends LockScope
{
	private final List<Customer> customers = new ArrayList<>();
	
	public void addCustomer(Customer c)
	{
		write(() -> {
			this.customers.add(c);
			Application.storageManager().store(this.customers);
		});
	}
	
	public void traverseCustomers(Consumer<Customer> consumer)
	{
		read(() -> this.customers.forEach(consumer));
	}
}
```

### CODE REVIEW

**Comment by Bios-Marcel** [comment]
This is definitely useful, we did something like that too.

However, shouldn't this be part of the store, instead of the serializer?

File: base/src/main/java/org/eclipse/serializer/concurrency/StripeLockScope.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+{
+	private final transient StripeLockedExecutor executor = StripeLockedExecutor.New(this.stripeCount());
+	
+	protected StripeLockScope()
+	{
+		super();
+	}
+	
+	/**
+	 * Gets the maxiimum number of stripes used for the {@link StripeLockedExecutor}.
```
Comment (hg-ms) [inline base/src/main/java/org/eclipse/serializer/concurrency/StripeLockScope.java:36]: typo: should be maximum

File: base/src/main/java/org/eclipse/serializer/concurrency/StripeLockedExecutor.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	{
+		return new StripeLockedExecutor.Default(
+			positive(stripeCount)
+		);
+	}
+	
+	
+	public static class Default implements StripeLockedExecutor
+	{
+		private transient volatile ReentrantReadWriteLock[] reentrantLocks;
```
Comment (hg-ms) [inline base/src/main/java/org/eclipse/serializer/concurrency/StripeLockedExecutor.java:143]: field reentrantLocks may be final

Comment (fh-ms) [reply base/src/main/java/org/eclipse/serializer/concurrency/StripeLockedExecutor.java:143]: field can't be volatile and final at the same time, volatile is needed for double-checked locking

File: base/src/main/java/org/eclipse/serializer/concurrency/StripeLockScope.java (hunk 1 above)
Comment (fh-ms) [reply base/src/main/java/org/eclipse/serializer/concurrency/StripeLockScope.java]: fixed

**Comment by fh-ms** [comment]
> This is definitely useful, we did something like that too.
> 
> However, shouldn't this be part of the store, instead of the serializer? Just wondering, maybe the question is dumb cause I missunderstood something :D

It's part of the serializer's base module like all other common utilities, which could be useful for any other module.

**Review by zdenek-jonas — APPROVED** [review]

**Review by hg-ms — APPROVED** [review]

**Pull request merged**